Add multi-lora support for Triton vLLM backend

SamuraiBUPT opened the pull request:
Hello, the vllm recently support multi-lora feature, see this [PR](https://github.com/vllm-project/vllm/pull/1804) for more information. 

To support request for lora, I implemented the code for lora request in triton backend.

---

__The vLLM scheme__

Now (2023.11.29) vllm supports loading local weights to model backbone, (no auto-download from S3/Hugging Face hub), they claimed that they will implement the remote-weights downloading later. 

Besides, the LoRA path should be configured from sender, for more you can see there [interface](https://github.com/Yard1/vllm/blob/dd1726f61d1b356fafca130458ae869fb05665e4/vllm/lora/request.py#L5). 

---

__Triton scheme__

Under the circumstances, the triton should act as a transit point for user's request processing and converting.

For users, all they need to do is to provide a target lora name for querying:

```bash 
python3 client_lora.py -l alpaca
```

Then Triton will check if the target lora is in the `local lora weights` or not. If true, Triton will wrap this target lora information to a new object `LoRARequest` and send to AsyncEngine to execute.

See `model.py` for more information.

---

__Changes to model repository__

The `model.json` needs new args such as `"enable_lora": "true"` to support multi-lora in vllm, see [EngineArgs](https://github.com/Yard1/vllm/blob/dd1726f61d1b356fafca130458ae869fb05665e4/vllm/engine/arg_utils.py#L11) for more.

The `multi_lora.json` file was created to figure out the specific lora name for each lora adapter. `model.json` stores the information of vllm engine initialization, so information of lora weight path cannot be stored in that. I created a new file to manage all local lora weights.

It can be configured like this:

```json
{
    "alpaca": "/vllm_workspace/weights/loras/alpaca-lora-7b",
    "bactrian": "/vllm_workspace/weights/loras/bactrian-x-llama-7b-lora"
}
```

Triton backend will read this file when initializing as a repository for local lora weights.

---

__Test__

The code passed the tests with and without lora support. Not setting `enable_lora` does not affect the original code.

`python3 client_lora.py`

```
Hello, my name is
I am a 20 year old student from the Netherlands. I am currently
```

`python3 client_lora.py -l alpaca`

```
Hello, my name is
I am a student at
I am currently studying
I am interested in learning
```

---

For more information about deployment, you can see the `docs/llama_multi_lora_tutorial.md`.

oandreeva-nv commented on src/model.py:
```diff
@@ -28,16 +28,19 @@
 import numpy as np
 import triton_python_backend_utils as pb_utils
-from vllm import SamplingParams
+from vllm.sampling_params import SamplingParams     # MODIFY: bug fix, it won't bother
```
could you please clarify this bug?

SamuraiBUPT replied:
I'm using `pip install -e .` to install vLLM from their source code (to test multi-lora version), in this case, using `from vllm import SamplingParams` will throw `ImportError`: 

```
ImportError: cannot import name 'SamplingParams' from 'vllm' (unknown location)
```

So I just tried to import the `SamplingParams` from the source code directly, to avoid this problem.

---

I think this modification will not affect both common users and developers, and solve the problem caused by `ImportError`.

oandreeva-nv replied:
Sounds good. I'll keep this conversation unresolved for history

oandreeva-nv commented on src/model.py:
```diff
@@ -91,7 +94,7 @@ def auto_complete_config(auto_complete_model_config):
     def initialize(self, args):
         self.logger = pb_utils.Logger
-        self.model_config = json.loads(args["model_config"])
+        self.model_config = json.loads(args["model_config"])    # config.pbtxt
```
I think we can remove this comment

SamuraiBUPT replied:
removed.

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,201 @@
+    └── config.pbtxt
+```
+
+Now, you have finish the basic deployment, and the file path should look like this:
````
```suggestion
Now, you have finished the basic deployment, and the file path should look like this:
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,201 @@
+
+
+
+__2.1 download the model repository files__
```
```suggestion
__2.1 Download the model repository files__
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,201 @@
+
+
+
+__2.2 complete the `model.json` according to your needs__
```
```suggestion
__2.2 Populate `model.json`__
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,201 @@
+
+__2.2 complete the `model.json` according to your needs__
+
+The content of `model.json` can be:
```
```suggestion
For this tutorial we will use the following set of parameters, specified in the `model.json`.
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,201 @@
++ `enable_lora`: If you want to support vllm multi-lora, this should be configured and set `true`.
++ `max_lora_rank`: The maximum of LoRA rank of your lora adapter.
+
+For more information, you may read the source code of [EngineArgs](https://github.com/Yard1/vllm/blob/multi_lora/vllm/engine/arg_utils.py#L11) here.
```
```suggestion
The full set of parameters can be found [here](https://github.com/Yard1/vllm/blob/multi_lora/vllm/engine/arg_utils.py#L11).
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,201 @@
+
+
+
+__2.3 add `multi_lora.json` to figure out the local lora path__
```
```suggestion
__2.3 Specify local lora path__
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,201 @@
+
+sudo docker run --gpus all -it --net=host -p 8001:8001 --shm-size=12G \
+--ulimit memlock=-1 --ulimit stack=67108864 -v ${PWD}:/vllm_workspace \
+-w /vllm_workspace nvcr.io/nvidia/tritonserver:23.10-vllm-python-py3 \
```
```suggestion
-w /vllm_workspace nvcr.io/nvidia/tritonserver:<xx.yy>-vllm-python-py3 \
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,201 @@
+-w /vllm_workspace nvcr.io/nvidia/tritonserver:23.10-vllm-python-py3 \
+/bin/bash
+```
+
````
```suggestion
 <xx.yy> is the version of Triton that you want to use. Please note, that Triton's vLLM container has been introduced starting from 23.10 release, and `lora` support was added in [update with target release container] release
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,201 @@
+/bin/bash
+```
+
+And download the `model.py` script from github:
````
In case `src/model.py` is updated with `lora`-related changes, next steps are redundant. As an option, we can keep them, but mention that "For pre-[target release version] containers, please replace provided in the container `/opt/tritonserver/backends/vllm/model.py` with the most up to date version.

Plus, we need to mention at what version of vllm `lora` support was added, since 23.10 container was build with 0.2.0 I believe, or later. 23.12 container comes with vllm version 0.2.3

SamuraiBUPT replied:
The current vllm version is 0.2.6, and multi-lora PR (by Yard1) is not merged yet. (multi-lora feature will be added in vllm 0.2.7 or later)

Since 23.12 container integrates vllm 0.2.3, the multi-lora feature will not be added temporarily. (Maybe `24.yy` container will support vllm 0.2.7+ version)

SamuraiBUPT replied:
I have understood what you mean, docs updated.

oandreeva-nv reviewed (commented):
Amazing job on this PR! I've noticed that you have a lot o `MODIFY` comments in `model.py`, is this PR still in progress? Feel free to tag me, when it is done, I'll review

SamuraiBUPT commented:
@oandreeva-nv Thank you very much for your thorough review of my code and for providing me with detailed guidance and suggestions. 

I have made changes to the code and the docs based on your suggestions.

Note that vllm multi-lora feature may not come with container 23.12 ( -> vllm 0.2.3). I will keep working on multi-lora support on backend, and fix bug or offer new features.

SamuraiBUPT commented:
2023-12-29 update:

+ Yard1's multi-lora PR hasn't been merged to main branch of vLLM yet, so I edited the `llama_multi_lora_tutorial.md` to offer my way of installing vLLM with multi-lora feature.
    So more developers can have easier and clear ways to get on with it.
+ In the latest commit, I fix the bug of non-graceful terminate error, the original code **lacks something in request error handling**. When terminating the triton server after LoRA request error response, it will trigger segment fault:
```
E1229 07:55:42.707509 6674 main.cc:517] failed to stop server: Internal - Exit timeout expired. Exiting immediately.
root@airaiot-MS-7D25:/vllm_workspace# I1229 07:55:44.442169 6737 pb_stub.cc:1815]  Non-graceful termination detected. 
*** SIGSEGV received at time=1703836544 on cpu 4 ***
PC: @     0x7f7311871a84  (unknown)  (unknown)
    @     0x7f7311817520  (unknown)  (unknown)
[2023-12-29 07:55:44,479 E 6737 6737] logging.cc:361: *** SIGSEGV received at time=1703836544 on cpu 4 ***
[2023-12-29 07:55:44,479 E 6737 6737] logging.cc:361: PC: @     0x7f7311871a84  (unknown)  (unknown)
[2023-12-29 07:55:44,480 E 6737 6737] logging.cc:361:     @     0x7f7311817520  (unknown)  (unknown)
Fatal Python error: Segmentation fault

Stack (most recent call first):
  <no Python frame>
```

and after fixing, the server will shut down smoothly:

```
I1229 08:03:17.533748 7999 server.cc:331] Timeout 29: Found 1 live models and 0 in-flight non-inference requests
I1229 08:03:17.536050 7999 model.py:362] [vllm] Issuing finalize to vllm backend
I1229 08:03:18.533890 7999 server.cc:331] Timeout 28: Found 1 live models and 0 in-flight non-inference requests
I1229 08:03:19.534038 7999 server.cc:331] Timeout 27: Found 1 live models and 0 in-flight non-inference requests
I1229 08:03:19.536044 7999 model.py:190] [vllm] Shutdown complete
(RayWorkerVllm pid=9004) INFO 12-29 08:02:44 model_runner.py:567] Graph capturing finished in 35 secs.
(RayWorkerVllm pid=9004) [W CUDAGraph.cpp:145] Warning: Waiting for pending NCCL work to finish before starting graph capture. (function operator())
I1229 08:03:20.534161 7999 server.cc:331] Timeout 26: Found 1 live models and 0 in-flight non-inference requests
I1229 08:03:21.534340 7999 server.cc:331] Timeout 25: Found 1 live models and 0 in-flight non-inference requests
I1229 08:03:22.534539 7999 server.cc:331] Timeout 24: Found 1 live models and 0 in-flight non-inference requests
I1229 08:03:22.783541 7999 model_lifecycle.cc:603] successfully unloaded 'vllm_model' version 1
I1229 08:03:23.534724 7999 server.cc:331] Timeout 23: Found 0 live models and 0 in-flight non-inference requests
root@lee-MS-7D25:/vllm_workspace#
```

---

**The following part is an error report.**

The 88th line of `src/model.py`:

```python
auto_complete_model_config.set_model_transaction_policy(dict(decoupled=True))
```

will cause error when launching the triton server, here are the logs:

```
I1229 07:19:29.086536 1055 pb_stub.cc:257]  Failed to initialize Python stub for auto-complete: AttributeError: 'ModelConfig' object has no attribute 'set_model_transaction_policy'

At:
  /opt/tritonserver/backends/vllm/model.py(88): auto_complete_config

E1229 07:19:29.102646 1042 model_lifecycle.cc:621] failed to load 'vllm_model' version 1: Internal: AttributeError: 'ModelConfig' object has no attribute 'set_model_transaction_policy'

At:
  /opt/tritonserver/backends/vllm/model.py(88): auto_complete_config

I1229 07:19:29.102667 1042 model_lifecycle.cc:756] failed to load 'vllm_model'
```

In my practice, I simply commented out that line of code, and it actually works.

Here is my dev environment:

+ docker env: nvcr.io/nvidia/tritonserver:23.10-vllm-python-py3
+ vLLM: yard's vllm fork repository (**I think the problem was not caused by vllm, but by triton lib**)

---

Lastly, hope this PR can get merged. :) @oandreeva-nv

germanjke commented:
which model.py we should update? multi-lora feature
this one https://raw.githubusercontent.com/triton-inference-server/vllm_backend/main/src/model.py looks like have no multi-lora support

germanjke commented:
ok right now evertyhing is fine, if I clone main vllm last version, not some branch which you posted in docs

SamuraiBUPT commented:
Everything in this PR is current with the latest vLLM version. I've revised the documentation to include instructions on installing vLLM with punica kernel compilation.

oandreeva-nv commented on src/model.py:
```diff
@@ -116,7 +119,19 @@ def initialize(self, args):
 
+        if "enable_lora" in vllm_engine_config.keys() and vllm_engine_config["enable_lora"]:
+            with open(multi_lora_args_filepath) as lora_file:
+                lora_repository: Dict[str, str] = json.load(lora_file)  # lora_repository = {"lora_name": "lora_path"}
```
Please, get rid of the comment

oandreeva-nv commented on src/model.py:
```diff
@@ -116,7 +119,19 @@ def initialize(self, args):
         )
+        
+        # create Triton LoRA weights repository
+        multi_lora_args_filepath = os.path.join(
```
how about we move this in the body of `if` on line 128?

oandreeva-nv commented on src/model.py:
```diff
@@ -252,11 +267,24 @@ async def generate(self, request):
                 parameters = request.parameters()
 
             sampling_params_dict = self.get_sampling_params_dict(parameters)
+            
```
Please, get rid of this empty line

oandreeva-nv commented on src/model.py:
```diff
@@ -252,11 +267,24 @@ async def generate(self, request):
             sampling_params_dict = self.get_sampling_params_dict(parameters)
+            
+            lora_name = sampling_params_dict.pop("lora_name", None)
+
```
same here

oandreeva-nv commented on src/model.py:
```diff
@@ -252,11 +267,24 @@ async def generate(self, request):
             sampling_params = SamplingParams(**sampling_params_dict)
 
             last_output = None
+            
+            # create LoRARequest
+            if lora_name is not None:
+                lora_id = str(self.supported_loras.index(lora_name) + 1)
+                lora_int_id = int(lora_id)
+                lora_local_path = self.lora_repository[lora_name]
+                lora_request = LoRARequest(lora_id, lora_int_id, lora_local_path)
+            else:
+                lora_request = None
+            
```
```suggestion
            lora_request = None
            if lora_name is not None:
                lora_id = str(self.supported_loras.index(lora_name) + 1)
                lora_int_id = int(lora_id)
                lora_local_path = self.lora_repository[lora_name]
                lora_request = LoRARequest(lora_id, lora_int_id, lora_local_path)
```

oandreeva-nv commented on src/model.py:
```diff
@@ -28,16 +28,19 @@
 from vllm.engine.async_llm_engine import AsyncLLMEngine
 from vllm.utils import random_uuid
 
+from vllm.lora.request import LoRARequest
```
Please, move this up on line 39

oandreeva-nv commented on src/model.py:
```diff
@@ -307,7 +335,34 @@ def execute(self, requests):
         """
         for request in requests:
-            self.create_task(self.generate(request))
+            # lora check logic
```
Could you please make this comment more descriptive?

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,243 @@
+```bash
+sudo docker run --gpus all -it --net=host -p 8001:8001 --shm-size=12G \
+--ulimit memlock=-1 --ulimit stack=67108864 -v ${PWD}:/vllm_workspace \
+-w /vllm_workspace nvcr.io/nvidia/tritonserver:23.12-vllm-python-py3 \
````
```suggestion
-w /vllm_workspace nvcr.io/nvidia/tritonserver:<xx.yy>-vllm-python-py3 \
```

SamuraiBUPT replied:
I will just remove this bash command example and reserve the former one.

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,243 @@
+The following tutorial demonstrates how to deploy **a LLaMa model** with **multiple loras** on Triton Inference Server using the Triton's [Python-based](https://github.com/triton-inference-server/backend/blob/main/docs/python_based_backends.md#python-based-backends) [vLLM](https://github.com/triton-inference-server/vllm_backend/tree/main) backend.
+
+
+## Step 0: Prepare vLLM multi-lora version
```
May I ask you to start tutorial's indexing from 1 and not 0?

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,243 @@
+
+
+## Step 0: Prepare vLLM multi-lora version
+git clone vllm repository:
```
Can we do `pip install` ?

oandreeva-nv replied:
Or because of `punica kernels` we need to install from source>

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,243 @@
+```
+
+This may take you 5-10 mins.
+
+
+
````
I think 2 lines is enough

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,243 @@
+mkdir -p vllm_workspace/weights
+cd vllm_workspace
+```
+
+
+
````
ditto

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,243 @@
+
+
+## Step 2: prepare model repository
+
+
+
```
1 line is enough

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,243 @@
+        │   └── model.json
+        └── config.pbtxt
+```
+
+
+
````
1 line is enough

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,243 @@
++ `max_lora_rank`: The maximum of LoRA rank of your lora adapter.
+
+The full set of parameters can be found [here](https://github.com/Yard1/vllm/blob/multi_lora/vllm/engine/arg_utils.py#L11).
+
+
+
```
ditto

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,243 @@
+```
+
+The **key** should be the supported lora name, and the **value** should be the specific path in your machine.
+
+
+
````
2 lines is fine

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,243 @@
+```bash
+cp ./model.py /opt/tritonserver/backends/vllm/
+```
+
+
+
````
2 lines is fine

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,243 @@
+I1030 22:33:28.292879 1 http_server.cc:4497] Started HTTPService at 0.0.0.0:8000
+I1030 22:33:28.335154 1 http_server.cc:270] Started Metrics Service at 0.0.0.0:8002
+```
+
+
+
````
2 lines is fine

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,243 @@
+```bash
+python3 client_lora.py -l alpaca
+```
+
+
+
````
remove extra lines

oandreeva-nv reviewed (commented):
I think this PR is in better shape already. I have 2 asks. 1) Either delete `multi_lora.json` from this PR, or populate it with sample values, you use in docs. 2) Could you please write some tests.

I would like to update vLLM version for Triton container first, but 0.2.7 has issues with memory, when running 2 vLLM engines at the same time, because of this I need to re-write our accuracy test. Does v0.2.7 of vLLM work? At the same time, I would assume you also need `punica kernels`, or LoRA can work without them? If the latter is not true, I would like to avoid changing `src/model.py`.

SamuraiBUPT commented:
The following commits do the following things:

+ Remove the `multi_lora.json` file in `samples` directory, users should create by themselves.
+ Remove redundant empty lines in `docs/llama_multi_lora_tutorial.md`
+ Revise the `docs/llama_multi_lora_tutorial.md`: **Preparing a docker container should be the first thing**, and other operations should be done inside the container environment. It's a terrible logic fault and I apologize for that.

SamuraiBUPT commented:
@oandreeva-nv Sir, I found the reason of accuracy test failing. In vLLM v0.2.7 they enabled CUDA Graph when starting the `llm_engine`, thus this should take extra gpu memory space.

I observed that in original `accuracy_test.py` line 67: `"gpu_memory_utilization": 0.3`, this may not come up to date with v0.2.7 which enables CUDA Graph.

In my practice, I increased this value to `0.4`. (one gpu, RTX 3090 24G), and I passed the `accuracy_test`.

Hope this will help you solve your problem, and I'm now working on tests over `multi-lora` feature.

SamuraiBUPT commented:
@oandreeva-nv Sir, I prepared a new workspace for vLLM v0.2.7 test.

My environment: 
+ triton docker image (vllm-python3) 23.12
+ RTX 3090 * 2
+ vLLM v0.2.7 release (build from source)

And I ran the `ci` tests to see memory problems.

The vLLM release v0.2.7 enables the feature of **CUDA graph**, thus the engine will take more gpu memory space more than it used to take. So the configurations in `ci` tests may not work fine.

There are two ways to solve this problem:
+ increase `gpu_memory_utilization`.
+ set `"enforce_eager": "true"` in `model.json` to avoid using CUDA Graph.

---

My test results:

+ `Accuracy test`: I modified `accuracy_test.py` line 67: `"gpu_memory_utilization": 0.3` to `"gpu_memory_utilization": 0.4`, and passed.
+ `enable_stream test`: passed.
+ `request_cancellation test`: passed.
+ `vllm_backend test`: Haven't passed this test yet, but I observed that `vllm_two` instance will only take about 700M of gpu memory and the process exited then.(In theory, it would take at least another 8000M gpu memory and trigger `exit`) I think it may has something to do with triton, not vllm itself.

---

The vLLM v0.2.7 does not support multi-lora feature yet. (maybe come in the future version) 

Thus my code may be tested with the latest vllm, built from source code.

I will submit more test scripts.

oandreeva-nv commented:
Hi @SamuraiBUPT , could you please resolve merge conflicts? We've also updated vLLM version to 0.3.0

SamuraiBUPT commented:
Hello, @oandreeva-nv  conflicts have been resolved, and the code passed all ci tests.

oandreeva-nv commented on ci/L0_backend_vllm/test.sh:
```diff
@@ -30,6 +30,8 @@ SUBTESTS="accuracy_test request_cancellation enabled_stream vllm_backend"
 
 python3 -m pip install --upgrade pip && pip3 install tritonclient[grpc]
 
+cp ../../src/model.py /opt/tritonserver/backends/vllm/
```
Please, remove it. In our ci, we build container with `model.py` from a specified branch already present in `/opt/tritonserver/backends/vllm/`

oandreeva-nv commented on ci/L0_multi_gpu/test.sh:
```diff
@@ -40,6 +40,7 @@ EXPECTED_NUM_TESTS=1
 
 rm -rf models && mkdir -p models
 cp -r ${SAMPLE_MODELS_REPO}/vllm_model models/vllm_opt
+cp ../../src/model.py /opt/tritonserver/backends/vllm/
```
ditto

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,239 @@
+# Depolying multi-lora vLLM backend in Triton
+
+The idea of multi-lora was proposed recently, for more you can read the paper:
```
```suggestion
The idea of multi-lora was proposed recently, for more please refer to:
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,239 @@
+
+**NOTICE:** the version of triton docker image should be configurated, here we use `<xx.yy>` to symbolize.
+
+Triton's vLLM container has been introduced starting from 23.10 release, and `lora` support will be added in future release.
```
```suggestion
Triton's vLLM container has been introduced starting from 23.10 release, and `lora` experimental support was added in vLLM v0.3.0 release.
```
Feel free to update vLLM version to a correct one with LoRA

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,239 @@
+
+For **pre-24.yy containers**, the docker images don't support multi-lora feature, so you need to replace that provided in the container `/opt/tritonserver/backends/vllm/model.py` with the most up to date version. Just follow these steps:
+
+Download the `model.py` script from github:
+
+```bash
+wget -P model_repository/vllm_model/1 https://raw.githubusercontent.com/triton-inference-server/vllm_backend/main/src/model.py
+```
+
+Copy this script to the backend path of triton inside the container:
+
+```bash
+cp ./model.py /opt/tritonserver/backends/vllm/
+```
````
```suggestion
Download the `model.py` script from github:

```bash
wget -P /opt/tritonserver/backends/vllm/ https://raw.githubusercontent.com/triton-inference-server/vllm_backend/main/src/model.py
```

```

Maybe combining the 2 steps will be easier to undeerstand?

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,239 @@
+__2.1 install with pip__
+
+```bash
+VLLM_INSTALL_PUNICA_KERNELS=1 pip install vllm
````
here and later, please include reference to vllm version used by Triton for consistency

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,239 @@
+    └── bactrian-x-llama-lora-7b
+```
+
+> A workspace for `vllm`, and `model backbone weights`, `LoRA adapter weights` is strongly recommended. 
+> 
+> You should expand the storage of these weight files to ensure they are logically organized in the workspace.
````
```suggestion
A workspace for `vllm`, and `model backbone weights`, `LoRA adapter weights` is strongly recommended. 
You should expand the storage of these weight files to ensure they are logically organized in the workspace.
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,239 @@
+cd /vllm_workspace
+
+mkdir -p model_repository/vllm_model/1
+wget -P model_repository/vllm_model/1 https://raw.githubusercontent.com/triton-inference-server/vllm_backend/main/samples/model_repository/vllm_model/1/model.json
```
```suggestion
wget -P model_repository/vllm_model/1 https://raw.githubusercontent.com/triton-inference-server/vllm_backend/r<xx.yy>/samples/model_repository/vllm_model/1/model.json
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,239 @@
+
+mkdir -p model_repository/vllm_model/1
+wget -P model_repository/vllm_model/1 https://raw.githubusercontent.com/triton-inference-server/vllm_backend/main/samples/model_repository/vllm_model/1/model.json
+wget -P model_repository/vllm_model/ https://raw.githubusercontent.com/triton-inference-server/vllm_backend/main/samples/model_repository/vllm_model/config.pbtxt
```
```suggestion
wget -P model_repository/vllm_model/ https://raw.githubusercontent.com/triton-inference-server/vllm_backend/r<xx.yy>/samples/model_repository/vllm_model/config.pbtxt
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,239 @@
+
+__4.3 Specify local lora path__
+
+So far the [vLLM multi-lora PR](https://github.com/vllm-project/vllm/pull/1804) just supported the inference of **local lora weights applying**, which means that the vllm cannot pull any lora adapter from huggingface. So triton should know where the local lora weights are.
```
Can we add a version reference, please? Something like as of v0.3.0 or v0.3.2?

SamuraiBUPT commented:
Hello, I have added extra ci test scripts for multi-lora feature.

## Environment Requirements and Explanation

The `multi-lora` test requires more than two GPUs to support this new feature. 

*(My test env is RTX 3090 * 2 (24G VRAM each))*

This can be explained from two aspects:
+ There is a `multi_gpu` ci test in this repo, so I assume that your environment supports this requirement.
+ Considering the current ecosystem of Llama models and LoRA, the vast majority of LoRA adapters are closely related to Llama models. Thus I simply choose `llama-7b-hf` and its LoRA adapters for ci test.
    + The size of Llama-7b is more than 13G in GPU memory.
    + vLLM will take extra GPU memory when profiling, for its PagedAttention.
    + Multi LoRA feature requires at least 10 LoRA adapter slots, which also requires GPU memory space.

So all these things account for its requirement.

> By the way, if your GPU memory is more than 40G (like A100), just ignore these settings :)

## What exactly does this new CI test do?

I wrote this ci test for two purposes:
+ Test whether vllm supports different LoRA adapters simultaneously.
+ Test whether vllm handles unsupported LoRAs.
    + vLLM does not support pulling adapters dynamically from huggingface, so Triton acts as a port for managing local LoRA weights, thus Triton should refuse requests which querying non-existing LoRA adapters.

## How long might this new CI test run?

That depends. 

As I said it is tested over Llama-7b and its adapters. So before this test, all the weights will be downloaded from huggingface through `download.py`. The downloading may take a few minutes.

*Just need to download once because there is symbolic links*

And setting `"tensor_parallel_size": 2` may take longer than one gpu scenario.

## More changes

The `multi_lora.json` in `samples/model_repository/vllm_model/1/` directory was deleted.

Because I think this file should not exist in this official repository, and this file should be created and populated by developers themselves, based on their own LoRA adapter needs.

jonatankawalek commented:
Great work @SamuraiBUPT,

Eagerly awaiting this PR to be merged, as it opens up a whole suite of possibilities for my projects.

jonatankawalek commented:
@oandreeva-nv Is there a rough timeline for when this PR could be merged?

jorge-tromero commented:
This feature is a great addition to Triton 🙏

oandreeva-nv commented:
Hi @SamuraiBUPT, I've just returned after my OOTO. Will start an internal CI and go through this PR again

oandreeva-nv commented on src/model.py:
```diff
@@ -361,7 +385,37 @@ def execute(self, requests):
+                    output_tensors=[output_tensor], error=lora_error
+                )
+                response_sender = request.get_response_sender()
+                response_sender.send(response)
+                response_sender.send(flags=pb_utils.TRITONSERVER_RESPONSE_COMPLETE_FINAL)
```
```suggestion
                response_sender.send(response, flags=pb_utils.TRITONSERVER_RESPONSE_COMPLETE_FINAL)
```

oandreeva-nv commented on ci/L0_backend_vllm/multi_lora/test.sh:
```diff
@@ -0,0 +1,122 @@
+    "tensor_parallel_size": 2,
+    "block_size": 16,
+    "enforce_eager": "true",
+    "enable_lora": "true",
```
Can we also test the case when `enable_lora` is present and `false`, please?

SamuraiBUPT commented:
@oandreeva-nv Thank you for your suggestions! There have been many changes in model.py and the CI tests. 

Some logical flaws and hidden errors have identified in the code. Specifically, these issues include:

+ CI tests for the case where `enable_lora` is set to false.
+ Server logic for processing and filtering LoRa requests.
+ A vulnerability in identifying `enable_lora` from the `multi_lora.json` file.

---

Regarding your suggestion to "encapsulate this check in a helper function," it's important to note that the requests may not be easily checked in the following pattern:

```python
def execute(self, requests):
    if self.enable_lora:    
        requests = self.verify_loras(requests)
    for request in requests:
        self.create_task(self.generate(request))
    return None
```

This is because we need to consider scenarios where a request querying for LoRA arrives at a server with LoRA disabled. Thus, it's essential that all requests undergo a helper function to filter out:

+ Requests querying for unsupported loras
+ Requests querying for lora while server doesn't support lora (this will trigger a vllm fault).

like:

```python
def execute(self, requests):
    requests = self.verify_loras(requests)
    for request in requests:
        self.create_task(self.generate(request))
    return None
```

By implementing this approach, Triton can act as a security guard to prevent the program from crashing during serving.

oandreeva-nv commented on src/model.py:
```diff
@@ -349,6 +375,45 @@ async def generate(self, request):
+            
+    def verify_loras(self, requests):
+        verified_requests = []
+        for request in requests:
```
I'm not a fan of this for loop, since we'll loop through requests twice in this case. How about we only keep the body of this loop in a helper function, then in execute we'll use it as:
```python
def execute(self, requests):
    for request in requests:
        request = self.verify_loras(request)
        self.create_task(self.generate(request))
    return None
```

But then we'd need to handle a case when we got an error and sent an error already. In this case maybe we'd need to return a `None` from helper function and hadle it inside `execute`, i.e. :
```python
def execute(self, requests):
    for request in requests:
        request = self.verify_loras(request)
        if request is not None:
            self.create_task(self.generate(request))
    return None
```

oandreeva-nv commented on src/model.py:
```diff
@@ -349,6 +375,45 @@ async def generate(self, request):
+                
+            if lora_error is not None:
+                output_tensor = pb_utils.Tensor(
+                    "text_output", np.asarray(["N/A"], dtype=self.output_dtype)
```
should we send an actual error back, instead on "N/A" ?

SamuraiBUPT replied:
Apology for my ignorance, but in this case error info has been sent by response_sender:

```python
response_sender.send(
    response, flags=pb_utils.TRITONSERVER_RESPONSE_COMPLETE_FINAL
)
```

and we can receive that:

```
Encountered error while processing: LoRA wizard is not supported, we currently support ['alpaca', 'wizardLM']
```

An error has been sent. I don't know what you mean by modifying `N/A`. I have re-write this code to 

```python
output_tensor = pb_utils.Tensor(
    "text_output",
    np.asarray(["[Error] Unsupported LoRA."], dtype=self.output_dtype),
)
```

If I misunderstood your suggestion, please let me know.

oandreeva-nv replied:
Apologies for not being clear. Thanks for changing "N/A", This is nice

oandreeva-nv commented on src/model.py:
```diff
@@ -122,7 +124,24 @@ def initialize(self, args):
+                self.enable_lora = True
+            except FileNotFoundError:
+                raise FileNotFoundError(f"Triton backend cannot find {multi_lora_args_filepath}.")
+            
```
```suggestion
```

oandreeva-nv commented on ci/L0_backend_vllm/multi_lora/download.py:
```diff
@@ -0,0 +1,47 @@
+# Copyright 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
```suggestion
# Copyright 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```

oandreeva-nv commented on ci/L0_backend_vllm/multi_lora/multi_lora_test.py:
```diff
@@ -0,0 +1,182 @@
+# Copyright 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
```suggestion
# Copyright 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```

oandreeva-nv commented on ci/L0_backend_vllm/multi_lora/test.sh:
```diff
@@ -0,0 +1,169 @@
+#!/bin/bash
+# Copyright 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
```suggestion
# Copyright 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,243 @@
+
+Triton's vLLM container has been introduced starting from 23.10 release, and `multi-lora` experimental support was added in vLLM v0.3.0 release.
+
+> Docker image version `nvcr.io/nvidia/tritonserver:23.12-vllm-python-py3` or higher version is strongly recommended.
```
```suggestion
> Docker image version `nvcr.io/nvidia/tritonserver:24.03-vllm-python-py3` or higher version is strongly recommended.
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,243 @@
+
+__2.1 install with pip__
+
+For Triton version before 24.02, you need the following command:
```
```suggestion
For Triton version before 24.05, you need the following command:
```

oandreeva-nv replied:
This feature will likely be released in 24.05 version. I'll update our vLLM installation to use punica kernels as well

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,243 @@
+VLLM_INSTALL_PUNICA_KERNELS=1 pip install vllm==0.3.0
+```
+
+__2.2 build from source (Recommended)__
````
Why is this option recommended?

SamuraiBUPT replied:
Sorry, it is a historical issue, when multi-lora feature hasn't been added to vLLM in early time, developers can only install vllm through source code to enable this feature. 

Now this tip can be removed :)

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,243 @@
+```bash
+git clone https://github.com/vllm-project/vllm.git
+```
+
````
```suggestion
```

oandreeva-nv commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,243 @@
+
+```bash
+cd vllm
+VLLM_INSTALL_PUNICA_KERNELS=1 pip install -e .
````
I think `-e` can be removed, as this will install vllm in editor mode, which is not necessary for users

oandreeva-nv commented on samples/client_lora.py:
```diff
@@ -0,0 +1,282 @@
+#!/usr/bin/env python3
+
+# Copyright 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
```suggestion
# Copyright 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```

oandreeva-nv commented on src/model.py:
```diff
@@ -28,16 +28,18 @@
 import os
 import threading
-from typing import AsyncGenerator
+from typing import AsyncGenerator, Dict, List
```
Could you please update copyright here:

2023 -> 2023-2024

oandreeva-nv commented:
@SamuraiBUPT , may I also ask to fix some pre-commit issues? 

The easiest way is to install  `pre-commit` : https://pre-commit.com/
and run `pre-commit run --all-files`

Some spelling issues should be fixed manually

SamuraiBUPT commented:
@oandreeva-nv Modifications have been made, and I have run pre-commit check locally and fix some issues as well.

Leave an issue about sending LoRA error.

oandreeva-nv commented:
Created pipeline (id: 14152157 )

oandreeva-nv commented:
Hi @SamuraiBUPT, may I ask to re-base your PR on top of a main branch? We've recently updated vllm version to 0.4.0 and it required some test changes

SamuraiBUPT commented:
@oandreeva-nv done :)

oandreeva-nv commented:
re-running CI

SamuraiBUPT commented:
I apologize for my ignorance; I did not fully adhere to the contribution process outlined in the Triton Contribution Guidelines initially. 

However, I have now submitted a CLA to [triton-cla@nvidia.com](mailto:triton-cla@nvidia.com).

tanmayv25 commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,242 @@
+
+---
+
+For **pre-24.03 containers**, the docker images didn't support multi-lora feature, so you need to replace that provided in the container `/opt/tritonserver/backends/vllm/model.py` with the most up to date version. Just follow this command:
```
I believe this version(24.03) needs to updated to a version when this PR gets merged.  We have already code-frozen for 24.04 release. So, this would be 24.05 assuming PR gets merged before 24.05 code freeze. 
Can you change this version to 24.05 and add a TODO comment to check for the correct version?

SamuraiBUPT replied:
changed to 24.05, and added a TODO comment.

tanmayv25 commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,242 @@
+
+**Notice:** `r<xx.yy>` is the triton version you need to configure.
+
+This command will download the `model.py` script to the Triton vllm backend directory to support multi-lora feature.
```
```suggestion
This command will download the `model.py` script to the Triton vllm backend directory which will enable multi-lora feature.
```

tanmayv25 commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,242 @@
+wget -P /opt/tritonserver/backends/vllm/ https://raw.githubusercontent.com/triton-inference-server/vllm_backend/r<xx.yy>/src/model.py
+```
+
+**Notice:** `r<xx.yy>` is the triton version you need to configure.
````
```suggestion
**Notice:** `r<xx.yy>` is the triton version you need to configure to r24.04 or later release.
```

tanmayv25 commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,242 @@
+cd /vllm_workspace
+```
+
+**NOTICE**: To enable multi-lora feature and speed up the inference, developers have integrated punica kernels into the `csrc` directory. To compile the punica kernels, you need to turn the `VLLM_INSTALL_PUNICA_KERNELS` env variable on to allow punica kernels compilation.
````
```suggestion
**NOTICE**: To enable multi-lora feature and speed up the inference, vLLM has integrated punica kernels. To compile the punica kernels, you need to turn the `VLLM_INSTALL_PUNICA_KERNELS` env variable on to allow punica kernels compilation.
```

tanmayv25 commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,242 @@
+VLLM_INSTALL_PUNICA_KERNELS=1 pip install .
+```
+
+This may take you 5-10 mins.
````
If it takes only 5-10 min, do you think we can enable VLLM_INSTALL_PUNICA_KERNELS  by default in our container image? The users would then not have to go through this extra step and can utilize LoRA directly. 
However, I see this issue which complains that it takes significantly longer to build the kernels.
https://github.com/vllm-project/vllm/issues/2604

oandreeva-nv replied:
Was also intended to ask that. I don't see any issues with us doing it in our container as a follow-up to this PR. Can come back and update these docs after the change

tanmayv25 commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,242 @@
+
+__4.3 Specify local lora path__
+
+vLLM v0.3.0 just supported the inference of **local lora weights applying**, which means that the vllm cannot pull any lora adapter from huggingface. So triton should know where the local lora weights are.
```
Has this changed for vLLM 0.4.0? Triton has advanced the vLLM version to 0.4.0.post1. See here: https://github.com/triton-inference-server/server/blob/main/build.py#L79
I believe we should mention same vLLM version that we are shipping with Triton 24.05. 

@oandreeva-nv  Thoughts?

oandreeva-nv replied:
Let's confirm the behavior in vLLM 0.4.0.post1.

tanmayv25 commented on docs/llama_multi_lora_tutorial.md:
```diff
@@ -0,0 +1,242 @@
+
+## Step 6: Send a request
+
+A client request script for multi-lora was prepared, downloading the client script from source:
```
A client request script for multi-lora is included in the samples: [client_lora.py](../samples/client_lora.py)

tanmayv25 commented on docs/llama_multi_lora_tutorial.md:
````diff
@@ -0,0 +1,242 @@
+
+A client request script for multi-lora was prepared, downloading the client script from source:
+
+```bash
````
Remove this... The above relative link will point to the client for the specific release. There is a possibility that there might be some breaking changes needed in the client script and the script from main might not work with older release.

tanmayv25 commented on ci/L0_backend_vllm/test.sh:
Update copyright years.

tanmayv25 commented on docs/llama_multi_lora_tutorial.md:
Add copyright here.

tanmayv25 commented on samples/client_lora.py:
Remove this file and add the `--lora-name` option in the current samples/client.py.
In the options documentation, you can mention that this option should be use iff using multi-lora in vLLM.

Maintaining two identical copies of client cause a lot of repeated code.

SamuraiBUPT replied:
`client_lora.py` was removed, and added `--lora-name` to `client.py`, it works fine now.

oandreeva-nv commented on ci/L0_backend_vllm/multi_lora/test.sh:
```diff
@@ -0,0 +1,169 @@
+pip install -U huggingface_hub
+
+rm -rf weights && mkdir -p weights/loras/alpaca && mkdir -p weights/loras/WizardLM
+mkdir -p weights/backbone/llama-7b-hf
```
Could you please recommend or try any smaller models to test with?

To make it easier for CI runners and not occupy much space. Otherwise we'll have a bunch of false negative tests due to `No space left` on the device, which I'm currently facing: CI 14199572

SamuraiBUPT replied:
changing to gemma-2b with its relative adapters

SamuraiBUPT commented:
@oandreeva-nv @tanmayv25 Hello, I have addressed the following issues in the new commit:

1. Updated copyright comments in the documentation and the copyright year in test scripts.
2. Revised the tutorial document to clarify `NGC version`, `vllm version`, and `Step 6: send a request`, along with some wording enhancements.
3. Modified the CI test scripts to use `gemma-2b` and its two related LoRA adapters, aligning the CI test code with current requirements.
4. Removed `client_lora.py` to eliminate redundant code.

---

Regarding the other issues:

- **Default installation of VLLM_INSTALL_PUNICA_KERNELS in container images:** I recommend enabling it by default. I have uninstalled vllm v0.3.0 and installed vllm v0.4.0.post1 from source, which took approximately 10 minutes and 31 seconds on a 12th Gen Intel(R) Core(TM) i9-12900K. The installation time might vary for other developers.
- **Changes to `local lora weights applying`?** No changes in v0.4.0.post1; it does not support dynamic adapter weight fetching from Hugging Face. If there are updates to the weight loading process, please inform me.
- **Transitioning CI tests from llama to gemma:** Understanding the space constraints in our CI test environment, I have reviewed all supported models and LoRAs as listed in the vllm documentation [Supported Models](https://github.com/vllm-project/vllm/blob/main/docs/source/models/supported_models.rst). `Gemma-2b` is the smallest model (**compatible with LoRA**) I can find. I have opted for this model hoping it aligns well with our CI requirements.

*By the way, I don't know where our pipeline is. I greatly appreciate you providing the pipeline ID, but could you also provide the location of the pipeline, or the platform it runs on?*

oandreeva-nv commented:
Hi @SamuraiBUPT , thanks for the thorough review! I'll re-run the pipeline and let you know regarding results.

At the moment CI/CD process is internal and external contributors do not have an access to it, thus maintainers of the Triton codebase are responsible for making sure CI/CD has passed for all external PRs.

I've started a new CI pipeline: 14260916.

tanmayv25 commented:
@SamuraiBUPT Thanks for resolving my questions. The code changes looks good to me. I will defer to @oandreeva-nv  for final approval. 

We will enable the PUNICA kernels in our container image as a follow-up item.

SamuraiBUPT commented:
HI @oandreeva-nv could you please let me know how it's going? If there are any issues with the CI test results that require bug fixes, I can help resolve them. I appreciate your time. :)

oandreeva-nv commented:
tests passed on A100: jobs/89630242 
I'll merge this PR today. Thank you @SamuraiBUPT for the amazing work and this valuable contribution!

oandreeva-nv reviewed (approved):
Amazing work!
